### ISSUE TITLE
Docker container running puppeteer connecting to headless chrome docker container?

### ISSUE THREAD

**Issue by saumets** [puppeteer/puppeteer#1645]
Have a Docker setup where:
* Container 1 -  is running headless Chrome
* Container 2 - runs NodeJS app.js w/ Puppeteer to control Chrome in Container 1.

My Docker compose file:
```
version: '3'
services:
  chrome:
    build: ./chrome
    shm_size: '1gb'
    cap_add:
      - SYS_ADMIN
    ports:
      - '9222:9222'
  app:
    build: ./app
    links:
      - chrome
```

Using the following code in app.js to connect Puppeteer and connect to https://www.google.com. Am making an initial request to `/json/version` to obtain the `webSocketDebuggerUrl` which Puppeteer will use:

```
const puppeteer = require('puppeteer');
const request = require('request-promise-native');

const options = {
	uri: `http://chrome:9222/json/version`,
	json: true,
	resolveWithFullResponse: true
};

request(options)
  .then((res) => {
	let webSocket = res.body.webSocketDebuggerUrl;
	console.log(`WebsocketUrl: ${webSocket}`);

	(async () => {
	   try {
      	      const browser = await puppeteer.connect({browserWSEndpoint: webSocket});
      	      const page    = await browser.newPage();
	      await page.goto(`https://wwww.google.com`, {waitUntil: 'networkidle2'});
	   }
	   catch(e) {
	      console.log(e);
	   }
    })();
});
```

I get the following when I run everything:
```
chrome_1  |
chrome_1  | DevTools listening on ws://0.0.0.0:9222/devtools/browser/d3d57807-1653-4292-aeb9-1b365e899ad7
chrome_1  | [1221/180304.301034:ERROR:gpu_process_transport_factory.cc(1046)] Lost UI shared context.
app_1     |
app_1     | > puppet-app@0.0.1 start /app
app_1     | > node index.js
app_1     |
app_1     | WebsocketUrl: ws://chrome:9222/devtools/browser/d3d57807-1653-4292-aeb9-1b365e899ad7
app_1     | Error: net::ERR_NAME_NOT_RESOLVED
app_1     |     at navigate (/app/node_modules/puppeteer/lib/Page.js:488:37)
app_1     |     at <anonymous>
app_1     |     at process._tickCallback (internal/process/next_tick.js:188:7)
```

`Error: net::ERR_NAME_NOT_RESOLVED` ?

Would anyone happen to know what the issue may be? Does Puppeteer not like the `chrome` hostname as part of the webSocketDebuggerUrl? I've tried replacing `chrome` in the webSocketDebuggerUrl with the actual IP of the chrome container but get the same results 😞 

Anyone run into this issue? Any help would be appreciated.

**Comment by saumets** [comment]
CLOSED. Typo: `page.goto("https://wwww.google.com", {waitUntil: 'networkidle2'});`. 😠

**Comment by rpascal** [comment]
@skalfyfan would you be willing to share your code on this project? I'm interested to see how you have it all set up because I am having difficulties getting headless chrome to run in is own container.

**Comment by saumets** [comment]
@rpascal sure. here you go:

Dockerfile for headless chrome:
```
FROM krallin/ubuntu-tini:xenial

EXPOSE 9222

# Install dependencies
RUN apt-get update && \
  apt-get -y upgrade && \
  apt-get install -yq curl libgconf-2-4

# Install Google Chrome
RUN curl https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - && \
  echo 'deb [arch=amd64] http://dl.google.com/linux/chrome/deb stable main' >> /etc/apt/sources.list.d/google-chrome.list && \
  apt-get update && \
  apt-get install -y google-chrome-unstable --no-install-recommends && \
  rm -fr /var/lib/apt/lists/* && \
  apt-get purge --auto-remove -y curl && \
  rm -fr /src/*.deb

COPY entrypoint.sh .
RUN chmod +x ./entrypoint.sh

ENTRYPOINT ["/usr/local/bin/tini", "--", "./entrypoint.sh"]
```

`entrypoint.sh` used for above Dockerfile:
```
#!/bin/bash

nohup google-chrome \
	--no-sandbox \
	--disable-background-networking \
	--disable-default-apps \
	--disable-extensions \
	--disable-sync \
	--disable-translate \
	--headless \
	--hide-scrollbars \
	--metrics-recording-only \
	--mute-audio \
	--no-first-run \
	--safebrowsing-disable-auto-update \
	--ignore-certificate-errors \
	--ignore-ssl-errors \
	--ignore-certificate-errors-spki-list \
	--user-data-dir=/tmp \
	--remote-debugging-port=9222 \
	--remote-debugging-address=0.0.0.0
```

My `app.js` which runs in it's own container which has nodeJS installed.
Note the line `http://chrome:9222/json/version` in the `options` object I create. Headless chrome is running in a container named `chrome`. I bring up the containers using _Docker Compose_ and name the service(container) `chrome` in there.

```
const puppeteer = require('puppeteer');
const request = require('request-promise-native');

const options = {
	uri: `http://chrome:9222/json/version`,
	json: true,
	resolveWithFullResponse: true
};

request(options)
  .then((res) => {
	let webSocket = res.body.webSocketDebuggerUrl;
	console.log(`WebsocketUrl: ${webSocket}`);

	(async () => {
	   try {
      	      const browser = await puppeteer.connect({browserWSEndpoint: webSocket});
      	      const page    = await browser.newPage();
	      await page.goto(`https://www.google.com`, {waitUntil: 'networkidle2'});
	   }
	   catch(e) {
	      console.log(e);
	   }
    })();
});
```

The above assumes that you want to run your NodeJS executing script in a Docker container as well. If you only want to run your headless chrome in a container and then locally run your app using Node then you can just set your application like you normally would and just change the URI in options from `http://chrome:9222/json/version` to uri: `http://localhost:9222/json/version` and then run your app via `npm start` locally. I have it working both ways. 

Hope that helps?

**Comment by rpascal** [comment]
That makes a lot of sense! This should be super helpful, thank you very much! I'm gonna try something like this tonight.

**Comment by drmrbrewer** [comment]
@skalfyfan thanks for sharing... and thanks for making a little typo in the first place because otherwise you wouldn't have shared ;-).

I'm really interested in this sort of approach, with a node app running in one container (in fact, several instances of the node app in multiple separate containers) and chrome in another.  Where is the chrome cache in your setup?  At present I set this with puppeteer via the `userDataDir` option, but this will create a cache for each instance.  Ideally I would like the central chrome container to hold the (single) cache, and the various node app containers calling the chrome container, with the central chrome cache being of benefit to all.

EDIT... I have tried this approach now, and it is working well (i.e. using a chrome instance in another container)... really neat... and it seems that the cache is also kept in the chrome container, in `/tmp`, because of the option in `entrypoint.sh`: `--user-data-dir=/tmp`.

I'd also like to work out how to do something similar for phantomjs, i.e. to have a single phantomjs instance in a container of its own, and to access phantomjs from multiple node apps in other containers via a similar mechanism... any hints?  At present I just have the phantomjs executable installed within same container as the node app(s) that use it, and call it directly using `execFile('phantomjs' ... )`.

**Comment by drmrbrewer** [comment]
@skalfyfan how would you deal with a crash of the browser in the 'chrome' container... if that happens then there is no chrome instance to connect to from the node app container?

**Comment by saumets** [comment]
@drmrbrewer glad to see you got your cache issue figured out! 👍 

RE: crash of browser in the 'chrome' container

If there was a browser crash I'd view that as a failed test and simply investigate. I guess it depends on your use-case for _puppeteer_ though. 🤔 

If I required what you're asking I'd probably look at docker container restart policies and get this setup in `docker-compose`: https://docs.docker.com/compose/compose-file/#restart_policy

In your Node app container I'd look at using something like `wait-for-port`:
https://www.npmjs.com/package/wait-for-port

The setup?

If your chrome container crashes, then your puppeteer orchestration should also crash in the node app container. Assuming container restart policies are set correctly then the chrome container should restart first, and the node app container should restart as well.

The `wait-for-port` module in the node app container should pause and wait for the chrome container to be ready before it restarts your puppeteer orchestration.

That should work I think (?) ..... in theory. I may spike it out myself to see...

**Comment by drmrbrewer** [comment]
@skalfyfan thanks for responding!  As for the use-case, yes I agree that if there is a crash then it needs investigating, but the idea is just to be resilient to such events when they happen, so that everything doesn't go belly up for hours before you notice!

My experience since posting this question is that occasionally my chrome container will encounter a problem, and at the moment it seems to be restarting itself... well not so much restarting itself but maybe more like stopping and then being automatically replaced by a new one (I have a docker swarm going).  I'm still new to docker so I'm still trying to figure it out :-)  At the moment I'm using the rather excellent [captainduckduck](https://github.com/githubsaturn/captainduckduck) to help me along and hold my hand.

Anyway, since the chrome container seems to re-incarnate itself, what I'm doing is listening for [the browser's `disconnected` event](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-disconnected) and using that as a trigger for connecting to the (re-incarnated) chrome service again (similar to your code above) to find the new `browserWSEndpoint` to use... if necessary I will retry connecting every 5 secs in case it hasn't yet started up again.  Probably your `wait-for-port` idea is more sophisticated than just "keep on trying".

Anyways, I'm not sure if that is best practice, but it does seem to be working OK.  Would be interested in your thoughts.  I also need to work out why chrome is crashing in the first place...

**Comment by MadHatter122312** [comment]
This is a fascinating example, but there is one thing I do not understand.  How exactly does your node app and the puppeteer script running in it 'connect' to Chrome Headless inside the container.  I  see that a websocket connection is opened to `http://chrome:9222/json/version`, then you goto() the url you wish to test like google.com.  
So, I assume that as your puppeteer script is running, as you are connected to Chrome Headless in the container and getting your response from chromium headless as json...Am I on the right track?

**Comment by drmrbrewer** [comment]
@IgorDraskovic I'm not the world's expert on this, and I am grateful to @skalfyfan for posting his example to teach how "chrome in a container" works, but my understanding is as follows.  The architecture is shown in the figure below (from [here](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#overview)).

![31592143-089f6f9a-b1db-11e7-9a20-16b7fc754fa1](https://user-images.githubusercontent.com/9009792/35739295-125baef0-0829-11e8-86d7-dfb793040db8.png)

As explained in the link, Puppeteer communicates with a browser using the devtools protocol.  Now, the link between puppeteer and the browser can be established by way of:

- [`puppeteer.launch()`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerlaunchoptions) or
- [`puppeteer.connect()`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerconnectoptions)

The former is when you want to launch a local instance of chrome, while the latter is when you want to connect to a remote instance (like in a Docker container).

However, in both cases you get a [`browser`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-browser), which acts in the same way regardless of which method you've used to create it.  So it doesn't much matter whether it's a local or remote chrome instance: you use the puppeteer API on the browser object in the same way and puppeteer does the tricky stuff for you.  

In other words, once you have your `browser` object, when you use it you don't "get a response from chromium headless as json" if you're using a remote instance... it works just the same as if you've launched a local instance and all the methods, like [`browser.newPage()`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#browsernewpage) etc work just the same.

That's about the limit of my understanding :-)

**Comment by MadHatter122312** [comment]
@drmrbrewer Oh, I think I understand...In my puppeteer scrip that runs with the bundled version of Chromium I have the following:

```
(async () => {

    const browser = await puppeteer.launch(
        {args: ['--no-sandbox', '--disable-setuid-sandbox'],
        ignoreHTTPSErrors: true 
    });
...
});
```
In the case of me refactoring my code to connect to Chromium running inside of a docker container I would have something like:

`const browser = await puppeteer.connect({browserWSEndpoint: webSocket});`

I guess what confuses me is that it seems that the entire test script should be inside of this `request(options)` block.  I assume this is because we need to maintain an open connection (websocket) to Chromium inside of the container while our test is running.

Is this correct?

**Comment by gzlock** [comment]
@skalfyfan Thank you for your sharing 🙏🙏🙏!
But how to let the docker container of chrome tell the ws url to the docker container of node.js, 
Is is possible?
Or let the chrome and the node.js into one container like  [docker/selenium/node-chrome](https://hub.docker.com/r/selenium/node-chrome/)?
BTW, I used docker-compose, the docker-compose.yml file:
```
version: "3"
services:
    mysql:
        ......
    redis:
        ......
    chrome:
        ......
    nodejs:
        ......
        links:
            - mysql
            - redis
            - chrome
    nginx:
        ......
        links:
            - nodejs
```

**Comment by saumets** [comment]
I've put a working example of this in a repo here:
https://github.com/skalfyfan/dockerized-puppeteer

I may put together a small article on it if time permits me 😄 

@gzlock You have to call the `/json/version` endpoint of the chrome debugger to obtain the websocket debugger url for Puppeteer to connect to so that it can orchestrate. This is documented here: https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerconnectoptions

Yes, alternatively you can run headless chrome and your node app in a single container. Then however you would be using `puppeteer.launch` as @drmrbrewer mentioned above.

**Comment by drmrbrewer** [comment]
Kudos, @skalfyfan!

Have you tried spinning up two instances of the chrome container?  I'm running a swarm and when my chrome service contains two chrome containers, I find that I can connect to the chrome service via `http://my-chrome-service` to get a valid response including `webSocketDebuggerUrl`, but then when I try to get a `browser` with `puppeteer.connect()` using that websocket, I get `Error: unexpected server response (404)`.

It's probably difficult for you to comment without seeing the source (it's a bit of a mess at present!) but maybe something immediately springs to mind.  I'm trying to see whether I can spread the load across multiple chrome containers (residing potentially on different servers in the docker swarm), but am not completely sure of the best approach :-/

**Comment by gzlock** [comment]
@drmrbrewer Maybe you can open multi chrome page to instead of multi chrome instance or the chrome docker container?

**Comment by drmrbrewer** [comment]
@gzlock at present I am using a single chrome instance (in the container) and am opening multiple pages as and when needed.  Whilst this works fine, when the page processing gets intensive, the CPU can max out (all pages link back to the same chrome instance on the same server).  So I'm working out whether it's possible to spread the load across different servers... i.e. a chrome container running on each server.

**Comment by joelgriffith** [comment]
This is pretty much the strategy I’ve employed on browserless here. The only difference is that this is a single container where WebSocket upgrades trigger a fresh Chrome launch, and the Node handler proxies the request into Chrome.

There’s a lot of other historical knowledge that’s now baked into the repo. It allows setting of launch flags via query-params, concurrency limits, and even queuing all set via environment variables. There’s even a debugger UI that comes with it to see what’s going on during your session.

Repo is here: https://github.com/joelgriffith/browserless. I’d love to see any issues you might have or how we can make it better.

**Comment by drmrbrewer** [comment]
@joelgriffith so how exactly would you suggest dealing with scaling up of the chrome instances?  I've got an open question on stackoverflow about this: https://stackoverflow.com/q/49028742/4070848.  I think you're suggesting that it's better not to have multiple chrome containers but instead to have a single chrome container running multiple chrome instances?  What if you need to spread the load across multiple servers, in docker swarm, to ease the load (even temporarily) on the existing server running the chrome container?  It seems that scaling by increasing container numbers would be a more flexible approach?

**Comment by joelgriffith** [comment]
@drmrbrewer answered your question on the stackoverflow link -- will follow up there with any subsequent questions

**Comment by cihadturhan** [comment]
**Update for future visitors**
You can now just use:
```js
const browser = await puppeteer.connect({
    browserURL: 'http://localhost:9222'
})
const page = await browser.newPage()
```
instead of sending a request for `websocketDebuggingUrl`.
Also, make sure you call
```js
await page.close();
```
when you are done. Otherwise, it will open a new tab everytime you run the script which will increase memory and idle cpu usage.
